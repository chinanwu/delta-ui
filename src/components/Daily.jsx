import FocusTrap from 'focus-trap-react';
import React, { lazy, useCallback, useEffect, useRef, useState } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Confetti from 'react-confetti';

import {
	aBtn,
	backspaceBtn,
	enterBtn,
	leftArrowBtn,
	rightArrowBtn,
	tabBtn,
	zBtn,
} from '../constants/Keycodes';
import { getFetch } from '../functions/FetchFunctions';
import formatCentisecondsTimer from '../functions/formatCentisecondsTimer';
import getThemeClassname from '../functions/getThemeClassname';
import isOneOff from '../functions/isOneOff';

const DailyWinModal = lazy(() => import('./DailyWinModal.jsx'));
import Loading from './Loading.jsx';
import HintButton from './HintButton.jsx';
import ThemeToggle from './ThemeToggle.jsx';

import './Daily.less';

const maxLen = 4;
const regex = /([a-zA-Z])/g;

export const Daily = ({ dark }) => {
	const [from, setFrom] = useState('');
	const [to, setTo] = useState('');
	const [date, setDate] = useState('');
	const [lowestHighscore, setLowestHighscore] = useState(-1);

	const [win, setWin] = useState(false);
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState('');

	const [numHints, setNumHints] = useState(3);
	const [hint, setHint] = useState(null);
	const [isHintExpanded, setIsHintExpanded] = useState(false);
	const [showHintInHistory, setShowHintInHistory] = useState(false);

	const [history, setHistory] = useState([]);
	const [guessVals, setGuessVals] = useState([]);
	const inputRefs = [useRef(null), useRef(null), useRef(null), useRef(null)];
	const historyBottomRef = useRef(null);

	const [timer, setTimer] = useState(0);

	const [score, setScore] = useState(0);

	useEffect(() => {
		let interval = null;
		if (!win) {
			interval = setInterval(() => {
				setTimer(seconds => seconds + 1);
			}, 10);
		}
		return () => clearInterval(interval);
	}, [timer, win]);

	useEffect(() => {
		document.title = 'Daily Challenge - Delta';
	}, []);

	useEffect(() => {
		setLoading(true);
		getFetch('/api/v1/dailychallenge')
			.then(res => {
				setDate(res.id);
				setFrom(res.from);
				setTo(res.to);
				setHistory([res.from]);
				setGuessVals(res.from.match(regex));
				setLowestHighscore(res.leaderboard[res.leaderboard.length - 1].score);
			})
			.then(() => setLoading(false));
	}, [setLoading, setDate, setFrom, setTo, setLowestHighscore]);

	useEffect(() => {
		historyBottomRef.current.scrollIntoView({
			behavior: 'smooth',
		});
	}, [history]);

	const handleHintClick = useCallback(() => {
		setLoading(true);
		getFetch(`/api/v1/hint?from=${history[history.length - 1]}&to=${to}`)
			.then(res => {
				if (res.hint && (res.numLeft === 0 || res.numLeft !== null)) {
					setNumHints(numHints => numHints - 1);
					// num left would be all words after last history word, includes to
					// e.g. heat -> cold, hint would be head, and numLeft 4 [head, held, hold, cold]
					setHint({ word: res.hint, numLeft: res.numLeft });
					setShowHintInHistory(true);
				} else {
					// Error is for guessing, need another way to log this
					// setError('Something went wrong grabbing game words!');
				}
			})
			.then(() => setLoading(false));
	}, [
		numHints,
		history,
		setNumHints,
		setHint,
		setShowHintInHistory,
		setLoading,
	]);

	const handleExpandHint = useCallback(
		expand => {
			setIsHintExpanded(expand);
		},
		[setIsHintExpanded]
	);

	const handleKeyDown = useCallback(
		event => {
			if (event && event.target && event.keyCode && event.key) {
				const i = parseInt(event.target.dataset.id);
				const next = (i + 1) % maxLen;
				const prev = i - 1 < 0 ? maxLen - 1 : i - 1;

				if (event.keyCode >= aBtn && event.keyCode <= zBtn) {
					event.preventDefault();
					guessVals[i] = event.key;
					inputRefs[i].current.value = event.key;
					if (next >= i) {
						inputRefs[next].current.focus();
						inputRefs[next].current.select();
					}
				} else {
					switch (event.keyCode) {
						case enterBtn:
							event.preventDefault();
							handleEnterClick();
							inputRefs[0].current.focus();
							break;
						case backspaceBtn:
							event.preventDefault();
							inputRefs[i].current.value = '';
							if (prev <= i) {
								inputRefs[prev].current.focus();
								inputRefs[prev].current.select();
							}
							break;
						case rightArrowBtn:
							inputRefs[next].current.focus();
							inputRefs[next].current.select();
							break;
						case leftArrowBtn:
							inputRefs[prev].current.focus();
							inputRefs[prev].current.select();
							break;
						case tabBtn:
							//do nothing
							break;
						default:
							event.preventDefault();
							break;
					}
				}
			}
		},
		[inputRefs]
	);

	const handleEnterClick = useCallback(() => {
		const guess = guessVals.join('');
		inputRefs.forEach(inputRef => {
			inputRef.current.value = '';
		});

		if (isOneOff(guess, history[history.length - 1])) {
			if (guess === to) {
				setLoading(true);
				getFetch(
					`/api/v1/score?from=${from}&to=${to}&time=${timer}&hintsUsed=${3 -
						numHints}&solution=${[...history, guess]}`
				)
					.then(res => {
						if (res) {
							setScore(res.score);
							setWin(true);
							setError(null);
							setHistory(history => [...history, guess]);
						}
					})
					.then(() => setLoading(false));
			} else {
				getFetch(`/api/v1/validate?word=${guess}`).then(res => {
					if (res.success) {
						setHistory(history => [...history, guess]);
						setError(null);
						setShowHintInHistory(false);
					} else {
						setGuessVals(history[history.length - 1].match(regex));
						setError("Word entered isn't a real word");
					}
				});
			}
		} else {
			setGuessVals(history[history.length - 1].match(regex));
			setError('Word must be one letter off from previous word');
		}
	}, [
		from,
		to,
		timer,
		numHints,
		guessVals,
		history,
		setScore,
		setWin,
		setError,
		setHistory,
		setLoading,
		setShowHintInHistory,
		setGuessVals,
		setTimer,
	]);

	const makeMysteryStepListItems = num => {
		let all = [];
		for (let i = 0; i < num; i++) {
			all[i] = (
				<li
					className="Daily__historyHintMysteryLi"
					key={`Daily__historyHintMystery--${i}`}
				>
					<div className="Daily__historyHintMystery">?</div>
				</li>
			);
		}
		return all;
	};

	return (
		<div className={getThemeClassname('Daily', dark)}>
			<ThemeToggle />

			<h2 className="Daily__header">Daily Challenge</h2>
			<p className="Daily__date">For the day of {date}</p>

			<h3 className="Daily__words">
				From: {from} -> To: {to}
			</h3>

			<div className="Daily__timer" role="timer">
				{formatCentisecondsTimer(timer)}
			</div>

			<div className={getThemeClassname('Daily__game', dark)}>
				<div className="Daily__history" aria-labelledby="dailyHistory">
					<h3 id="dailyHistory" className="Daily__historyLabel">
						History
					</h3>
					<ul
						className={
							getThemeClassname('Daily__historyList', dark) +
							(win ? ' Daily__historyList--hideScroll' : '')
						}
						aria-live="polite"
					>
						{history.map((item, i) => (
							<li key={`Daily__historyItem--${i}`}>{item}</li>
						))}
						<li ref={historyBottomRef} />

						{showHintInHistory && hint ? (
							hint.word !== to ? (
								<>
									<li
										className={getThemeClassname(
											'Daily__historyHintWord',
											dark
										)}
									>
										{hint.word}
									</li>
									{makeMysteryStepListItems(hint.numLeft - 2)}
									<li>{to}</li>
								</>
							) : (
								<>
									<li className="Daily__historyHintWord">{hint.word}</li>
								</>
							)
						) : null}
					</ul>
				</div>

				<div>
					<HintButton
						id="dailyHintBtn"
						numHints={numHints}
						isExpanded={isHintExpanded}
						dark={dark}
						btnText={`Get a Hint: ${numHints}`}
						ariaLabelledBy="dailyHintHeader"
						giveSolution={false}
						onClick={handleHintClick}
						onExpandChange={handleExpandHint}
					>
						<h3 id="dailyHintHeader">Hint: </h3>
						<p className="Daily__hint">
							Recommended next word: {hint ? hint.word : ''}
						</p>
						<p className="Daily__hint">
							Estimated number of words left: {hint ? hint.numLeft : ''}
						</p>
					</HintButton>

					<div className="Daily__guessContainer">
						<h3 id="dailyGuessLabel" className="Daily__guessLabel">
							Next Word:
						</h3>
						<div className="Daily__guessInputs">
							<div>
								{guessVals.map((val, i) => (
									<input
										id={'dailyGuessInput-' + i}
										ref={inputRefs[i]}
										data-id={i}
										className={getThemeClassname('Daily__guessInput', dark)}
										type="text"
										maxLength={1}
										placeholder={val}
										aria-labelledby="dailyGuessLabel"
										aria-invalid={error}
										onKeyDown={handleKeyDown}
										key={'Daily__guessInput-' + i}
									/>
								))}
							</div>

							<button
								id="dailyEnterBtn"
								className="Daily__enterBtn"
								onClick={handleEnterClick}
							>
								Enter
							</button>
						</div>

						<div className={getThemeClassname('Daily__gameError', dark)}>
							{error}
						</div>
					</div>
				</div>
			</div>

			{loading && createPortal(<Loading />, document.body)}
			{win &&
				createPortal(
					<FocusTrap>
						<div className="Daily__win">
							<Confetti number={50} recycle={false} />
							<DailyWinModal
								dark={dark}
								from={from}
								to={to}
								playerSoln={history}
								timer={timer}
								hintsUsed={3 - numHints}
								score={score}
								isHighscore={score > lowestHighscore}
							/>
						</div>
					</FocusTrap>,
					document.body
				)}
			<div className="Daily__easterEgg">This is an incredibly tiny screen</div>
		</div>
	);
};

Daily.propTypes = {
	dark: PropTypes.bool,
};

export const mapStateToProps = ({ theme: { dark } }) => ({
	dark,
});

export default connect(mapStateToProps)(Daily);

// TODO:
// - See Solo
