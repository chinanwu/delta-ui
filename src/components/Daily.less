@import '../style/colours';
@import '../style/button';

.Daily {
	height: 100%;
}

.Daily--dark {
	@dark();
}

.Daily__header {
	text-align: center;
	padding-bottom: 0;
}

.Daily__date {
	text-align: center;
	margin: 0;
}

.Daily__words {
	display: flex;
	justify-content: center;
	align-items: center;
	padding-top: 0;
	font-size: 2.4rem;
}

.Daily__hintContent {
	border: 1px solid black;
	display: flex;
	flex-direction: column;
	align-items: center;
	padding: 1rem;
}

.Daily__hintContent--dark {
	background-color: @dark-02;
	color: #fff;
}

.Daily__hint--short {
	margin-top: 0;
}

.Daily__enterBtn,
.Daily__hintCloseBtn {
	@smBtn();
}

.Daily__timer {
	text-align: center;
	font-size: 2.4rem;
}

.Daily__game {
	display: flex;
	justify-content: center;
	height: 70%;
	margin-top: 1rem;
}

.Daily__game--dark {
	@dark();
}

.Daily__history {
	display: flex;
	flex-direction: column;
	align-items: center;
	text-align: center; // For history label
}

.Daily__historyLabel {
	border: 1px solid @green;
	background-color: @green;
	width: 22rem;
	color: black;
}

.Daily__historyList {
	border: 1px solid black;
	list-style: none;
	padding: 1rem 0;
	margin: 0;
	overflow: scroll;
	width: 100%;
	height: 100%;
	font-size: 2.4rem;

	//Chrome/Edge/Safari
	&::-webkit-scrollbar,
	&::-webkit-scrollbar-track,
	&::-webkit-scrollbar-thumb {
		background-color: white;
	}

	//Firefox
	scrollbar-color: #b0b0b0 white; /* scroll thumb & track */
}

.Daily__historyList--dark {
	border: none;
	background-color: @dark-02;
	color: #fff;

	//Chrome/Edge/Safari
	&::-webkit-scrollbar,
	&::-webkit-scrollbar-track,
	&::-webkit-scrollbar-thumb {
		background-color: @dark-02;
	}

	//Firefox
	scrollbar-color: #323232 @dark-02; /* scroll thumb & track */
}

.Daily__historyList--hideScroll {
	overflow: hidden;
}

.Daily__historyHintWord {
	color: #757575;
}

.Daily__historyHintWord--dark {
	color: #aedb7a;
}

.Daily__historyHintMysteryLi {
	display: flex;
	justify-content: center;
}

.Daily__historyHintMystery {
	width: 10rem;
	margin-bottom: 0.5rem;
	text-align: center;
	background-color: black;
	color: white;
}

.Daily__historyHintMystery--dark {
	background-color: #3c2b3b;
	color: white;
}

.Daily__hint {
	font-size: 1.6rem;
	margin-bottom: 0.5rem;
	margin-top: 0;
}

.Daily__guessContainer {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	width: 30rem;
}

.Daily__guessLabel {
	margin-bottom: 2rem;
}

.Daily__guessInputs {
	display: flex;
	justify-content: center;
	position: absolute;
}

.Daily__guessInput {
	margin: 0.4rem;
	width: 3.6rem;
	height: 4.2rem;
	text-align: center;
	font-size: 1.6rem;

	&::-moz-placeholder {
		// Make placeholder text colour AAA
		color: #595959;
	}

	&:-ms-input-placeholder {
		color: #595959;
	}

	&::-webkit-input-placeholder {
		color: #595959;
	}
}

.Daily__guessInput--dark {
	color: white;
	background: @dark-02;
	border: 1px solid black;
	padding: 0.3rem;
	&::-moz-placeholder {
		// Make placeholder text colour AAA
		color: #bfbfbf;
	}

	&:-ms-input-placeholder {
		color: #bfbfbf;
	}

	&::-webkit-input-placeholder {
		color: #bfbfbf;
	}
}

.Daily__gameError {
	font-size: 1.8rem;
	text-align: center;
	color: @error;
	margin-top: 1rem;
	position: relative;
	top: 3rem;
	height: 6rem;
}

.Daily__gameError--dark {
	color: @errorDark;
}

.Daily__win {
	position: fixed;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
	height: 100%;
	width: 100%;
	display: flex;
	justify-content: center;
	align-items: center;
	flex-direction: column;
	color: @green;
	background-color: @modalBg;
}

.Daily__easterEgg {
	display: none;
}

@media only screen and (max-width: 550px) {
	.Daily__game {
		flex-direction: column;
		height: initial;
	}

	.Daily__enterBtn,
	.Daily__hint {
		font-size: 125%;
	}

	.Daily__hintContent {
		margin-right: 2rem;
		margin-left: 2rem;
	}

	.Daily__guessContainer {
		width: initial;
	}

	.Daily__guessLabel {
		margin-bottom: 0;
	}

	.Daily__guessInputs {
		position: initial;
	}

	.Daily__gameError {
		position: initial;
		top: initial;
	}

	.Daily__historyList {
		display: flex;
		align-items: center;
		flex-direction: column;
		font-size: 3rem;
		text-align: center;
		padding-left: 0;
		width: 22rem;
		height: 20rem;
	}

	.Daily__historyLabel {
		width: 20rem;
	}
}

@media only screen and (max-width: 375px) {
	.Daily__words {
		padding-top: 2rem;
	}

	.Daily__guessInput {
		font-size: 2rem;
	}
}

@media only screen and (max-width: 340px) {
	.Daily__header {
		margin-top: 4rem;
	}

	.Daily__words {
		text-align: center;
	}
}

@media only screen and (max-width: 310px) {
	.Daily__guessInputs {
		flex-direction: column;
	}
}

@media only screen and (max-width: 225px) {
	.Daily__easterEgg {
		display: initial;
		position: absolute;
		top: 0;
		font-size: 1rem;
	}
}
