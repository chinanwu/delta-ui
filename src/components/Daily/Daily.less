@import '../../style/colours';

.Daily {
	height: 100%;
}

.Daily--dark {
	@dark();
}

.Daily__header {
	text-align: center;
	padding-bottom: 0;
}

.Daily__date {
	text-align: center;
	margin: 0;
}

.Daily__words {
	display: flex;
	justify-content: center;
	align-items: center;
	padding-top: 0;
	font-size: 2.4rem;
}

.Daily__wordLine {
	display: flex;
	box-shadow: @greenLight 0px -1rem 0px inset;
	text-shadow: 0.05rem 0.05rem #fff, -0.05rem 0.05rem #fff,
		0.05rem -0.05rem #fff, -0.05rem -0.05rem #fff;

	transition: box-shadow 0.2s;
	&:hover {
		box-shadow: @greenLight 0px -4rem 0px inset;
	}
}

.Daily__wordLine--dark {
	box-shadow: @dark-10 0px -1rem 0px inset;
	text-shadow: none;

	transition: box-shadow 0.2s;
	&:hover {
		box-shadow: @dark-10 0px -4rem 0px inset;
	}
}

.Daily__wordLabel::after {
	content: ': ';
}

.Daily__wordArrow::before {
	content: '\00a0'
}

.Daily__wordArrow::after {
	content: '\00a0'
}

.Daily__timer {
	text-align: center;
	font-size: 2.4rem;
}

.Daily__game {
	display: flex;
	justify-content: center;
	height: 70%;
	margin-top: 1rem;
}

.Daily__game--dark {
	@dark();
}

.Daily__easterEgg {
	display: none;
}

@media only screen and (max-width: 550px) {
	.Daily__game {
		flex-direction: column;
		height: initial;
	}

	.Daily__wordLine {
		display: initial;
		text-align: center;
	}

	.Daily__wordLine {
		&:hover {
			box-shadow: @greenLight 0px -3rem 0px inset;
		}
	}

	.Daily__wordLine--dark {
		&:hover {
			box-shadow: @dark-10 0px -3rem 0px inset;
		}
	}
}

@media only screen and (max-width: 375px) {
	.Daily__words {
		padding-top: 2rem;
	}
}

@media only screen and (max-width: 340px) {
	.Daily__header {
		padding-top: 4rem;
	}

	.Daily__words {
		text-align: center;
	}
}

@media only screen and (max-width: 336px) {
	.Daily__wordLine {
		display: flex;
		flex-direction: column;
	}

	.Daily__wordLabel {
		font-size: 1.6rem;
	}

	.Daily__wordLabel:after {
		content: '';
	}
}

@media only screen and (max-width: 225px) {
	.Daily__easterEgg {
		display: initial;
		position: absolute;
		top: 0;
		font-size: 1rem;
	}
}