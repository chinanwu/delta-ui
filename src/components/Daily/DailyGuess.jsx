import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { applyGuess } from '../../thunk/DailyThunk.jsx';

import Guess from '../Guess.jsx';

export const DailyGuess = ({ prevWord, error, onGuess }) => {
	return <Guess prevWord={prevWord} error={error} onGuess={onGuess} />;
};

DailyGuess.propTypes = {
	prevWord: PropTypes.string,
	error: PropTypes.string,
	onGuess: PropTypes.func,
};

export const mapStateToProps = ({ daily: { prevWord, guessError } }) => ({
	prevWord,
	error: guessError,
});

const mapDispatchToProps = {
	onGuess: applyGuess,
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(DailyGuess);
