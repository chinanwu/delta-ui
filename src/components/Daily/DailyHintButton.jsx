import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { applyCloseHint, requestHint } from '../../thunk/DailyThunk.jsx';

import HintButton from '../HintButtonV2.jsx';

export const DailyHintButton = ({
	hint,
	numHints,
	isExpanded,
	error,
	onHint,
	onClose,
}) => (
	<HintButton
		id="dailyHintBtn"
		hint={hint}
		numHints={numHints}
		isExpanded={isExpanded}
		error={error}
		giveSolution={false}
		onHint={onHint}
		onClose={onClose}
	/>
);

DailyHintButton.propTypes = {
	hint: PropTypes.object,
	numHints: PropTypes.number,
	error: PropTypes.string,
	isExpanded: PropTypes.bool,
	onHint: PropTypes.func,
	onClose: PropTypes.func,
};

export const mapStateToProps = ({
	daily: { hint, numHints, hintExpanded },
}) => ({
	hint,
	numHints,
	isExpanded: hintExpanded,
});

const mapDispatchToProps = {
	onHint: requestHint,
	onClose: applyCloseHint,
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(DailyHintButton);
