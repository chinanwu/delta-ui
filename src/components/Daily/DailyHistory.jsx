import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import History from '../History.jsx';

export const DailyHistory = ({
	to,
	history,
	hintWord,
	hintNumLeft,
	win,
	showHint,
}) => (
	<History
		to={to}
		history={history}
		hintWord={hintWord}
		hintNumleft={hintNumLeft}
		win={win}
		showHint={showHint}
	/>
);

DailyHistory.propTypes = {
	to: PropTypes.string,
	history: PropTypes.arrayOf(PropTypes.string),
	hintWord: PropTypes.string,
	hintNumLeft: PropTypes.number,
	win: PropTypes.bool,
	showHint: PropTypes.bool,
};

export const mapStateToProps = ({
	daily: { to, history, hintWord, hintNumLeft, win, showHintInHistory },
}) => ({
	to,
	history,
	hintWord,
	hintNumLeft,
	win,
	showHint: showHintInHistory,
});

export default connect(mapStateToProps)(DailyHistory);
