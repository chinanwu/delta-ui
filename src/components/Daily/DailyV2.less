@import '../../style/colours';

.Daily {
	height: 100%;
}

.Daily--dark {
	@dark();
}

.Daily__header {
	text-align: center;
	padding-bottom: 0;
}

.Daily__date {
	text-align: center;
	margin: 0;
}

.Daily__words {
	display: flex;
	justify-content: center;
	align-items: center;
	padding-top: 0;
	font-size: 2.4rem;
}

.Daily__timer {
	text-align: center;
	font-size: 2.4rem;
}

.Daily__game {
	display: flex;
	justify-content: center;
	height: 70%;
	margin-top: 1rem;
}

.Daily__game--dark {
	@dark();
}

@media only screen and (max-width: 550px) {
	.Daily__game {
		flex-direction: column;
		height: initial;
	}
}

@media only screen and (max-width: 375px) {
	.Daily__words {
		padding-top: 2rem;
	}
}

@media only screen and (max-width: 340px) {
	.Daily__header {
		margin-top: 4rem;
	}

	.Daily__words {
		text-align: center;
	}
}
