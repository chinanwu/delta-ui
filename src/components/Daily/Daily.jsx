import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import { connect } from 'react-redux';

import formatCentisecondsTimer from '../../functions/formatCentisecondsTimer';
import getThemeClassname from '../../functions/getThemeClassname';
import { requestDailyChallenge } from '../../thunk/DailyThunk.jsx';

import Error from '../Error.jsx';
import withTitle from '../HOC/withTitle.jsx';
import Loading from '../Loading.jsx';
import ThemeToggle from '../ThemeToggle.jsx';

import DailyGuess from './DailyGuess.jsx';
import DailyHintButton from './DailyHintButton.jsx';
import DailyHistory from './DailyHistory.jsx';
import DailyWinModal from './DailyWinModal.jsx';

import './Daily.less';

export const Daily = ({
	dark,
	from,
	to,
	win,
	date,
	loading,
	error,
	getDaily,
}) => {
	const [timer, setTimer] = useState(0);

	useEffect(() => {
		let interval = null;
		if (!win) {
			interval = setInterval(() => {
				setTimer(seconds => seconds + 1);
			}, 10);
		}
		return () => clearInterval(interval);
	}, [timer, win]);

	useEffect(() => {
		if (!from || !to) {
			getDaily();
			setTimer(0);
		}
	}, [from, to, getDaily, setTimer]);

	return (
		<div className={getThemeClassname('Daily', dark)}>
			<ThemeToggle />

			<h2 className="Daily__header">Daily Challenge</h2>
			<p className="Daily__date">For the day of {date}</p>

			<h3 className="Daily__words">
				From: {from} -> To: {to}
			</h3>

			<div className="Daily__timer" role="timer">
				{formatCentisecondsTimer(timer)}
			</div>

			<div className={getThemeClassname('Daily__game', dark)}>
				<DailyHistory />

				<div>
					<DailyHintButton />
					<DailyGuess />
				</div>
			</div>

			<div className="Daily__easterEgg">This is an incredibly tiny screen</div>

			{loading && createPortal(<Loading />, document.body)}
			{error && createPortal(<Error dark={dark} />, document.body)}
			{win && createPortal(<DailyWinModal timer={timer} />, document.body)}
		</div>
	);
};

Daily.propTypes = {
	dark: PropTypes.bool,
	from: PropTypes.string,
	to: PropTypes.string,
	date: PropTypes.string,
	win: PropTypes.bool,
	timeStarted: PropTypes.number,
	error: PropTypes.object,
	loading: PropTypes.bool,
	getDaily: PropTypes.func,
};

export const mapStateToProps = ({
	daily: { from, to, date, win, error, loading },
	theme: { dark },
}) => ({
	dark,
	from,
	to,
	date,
	win,
	error,
	loading,
});

const mapDispatchToProps = {
	getDaily: requestDailyChallenge,
};

export default withTitle('Daily Challenge')(
	connect(
		mapStateToProps,
		mapDispatchToProps
	)(Daily)
);
