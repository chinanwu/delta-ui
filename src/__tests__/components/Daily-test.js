import { shallow } from 'enzyme';
import React, { useEffect } from 'react';

import { Daily, mapStateToProps } from '../../components/Daily.jsx';

jest.unmock('../../components/Daily.jsx');

describe('Daily', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<Daily />);
			expect(wrapper.getElement()).toMatchSnapshot();
			useEffect.mock.calls[0][0]();
		});

		it('renders Daily dark', () => {
			const wrapper = shallow(<Daily dark={true} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders Daily light', () => {
			const wrapper = shallow(<Daily dark={false} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	// describe('behaviour', () => {
	// 	it('sets the timer', () => {
	// useState
	// 	.mockReturnValueOnce(['#ffffff', setTest]);
	// const wrapper = shallow(<Daily />);
	// expect(wrapper.getElement()).toMatchSnapshot();
	// });
	// });

	describe('mapStateToProps', () => {
		it('returns mapped properties', () => {
			const theme = { dark: true };
			const state = { theme };
			expect(mapStateToProps(state)).toEqual(theme);
		});
	});
});
