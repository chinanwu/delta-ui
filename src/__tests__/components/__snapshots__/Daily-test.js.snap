// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Daily rendering renders Daily dark 1`] = `
<div>
  <Memo(Connect(ThemeToggle)) />
  <h2
    className="Daily__header"
  >
    Daily Challenge
  </h2>
  <p
    className="Daily__date"
  >
    For the day of 
    
  </p>
  <h3
    className="Daily__words"
  >
    From: 
    
     -&gt; To: 
    
  </h3>
  <div
    className="Daily__timer"
    role="timer"
  />
  <div>
    <div
      aria-labelledby="dailyHistory"
      className="Daily__history"
    >
      <h3
        className="Daily__historyLabel"
        id="dailyHistory"
      >
        History
      </h3>
      <ul
        aria-live="polite"
        className="undefined"
      >
        <li />
      </ul>
    </div>
    <div>
      <HintButton
        ariaLabelledBy="dailyHintHeader"
        btnText="Get a Hint: 3"
        dark={true}
        giveSolution={false}
        id="dailyHintBtn"
        isExpanded={false}
        numHints={3}
        onClick={[Function]}
        onExpandChange={[Function]}
      >
        <h3
          id="dailyHintHeader"
        >
          Hint: 
        </h3>
        <p
          className="Daily__hint"
        >
          Recommended next word: 
          
        </p>
        <p
          className="Daily__hint"
        >
          Estimated number of words left: 
          
        </p>
      </HintButton>
      <div
        className="Daily__guessContainer"
      >
        <h3
          className="Daily__guessLabel"
          id="dailyGuessLabel"
        >
          Next Word:
        </h3>
        <div
          className="Daily__guessInputs"
        >
          <div />
          <button
            className="Daily__enterBtn"
            id="dailyEnterBtn"
            onClick={[Function]}
          >
            Enter
          </button>
        </div>
        <div>
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="Daily__easterEgg"
  >
    This is an incredibly tiny screen
  </div>
</div>
`;

exports[`Daily rendering renders Daily light 1`] = `
<div>
  <Memo(Connect(ThemeToggle)) />
  <h2
    className="Daily__header"
  >
    Daily Challenge
  </h2>
  <p
    className="Daily__date"
  >
    For the day of 
    
  </p>
  <h3
    className="Daily__words"
  >
    From: 
    
     -&gt; To: 
    
  </h3>
  <div
    className="Daily__timer"
    role="timer"
  />
  <div>
    <div
      aria-labelledby="dailyHistory"
      className="Daily__history"
    >
      <h3
        className="Daily__historyLabel"
        id="dailyHistory"
      >
        History
      </h3>
      <ul
        aria-live="polite"
        className="undefined"
      >
        <li />
      </ul>
    </div>
    <div>
      <HintButton
        ariaLabelledBy="dailyHintHeader"
        btnText="Get a Hint: 3"
        dark={false}
        giveSolution={false}
        id="dailyHintBtn"
        isExpanded={false}
        numHints={3}
        onClick={[Function]}
        onExpandChange={[Function]}
      >
        <h3
          id="dailyHintHeader"
        >
          Hint: 
        </h3>
        <p
          className="Daily__hint"
        >
          Recommended next word: 
          
        </p>
        <p
          className="Daily__hint"
        >
          Estimated number of words left: 
          
        </p>
      </HintButton>
      <div
        className="Daily__guessContainer"
      >
        <h3
          className="Daily__guessLabel"
          id="dailyGuessLabel"
        >
          Next Word:
        </h3>
        <div
          className="Daily__guessInputs"
        >
          <div />
          <button
            className="Daily__enterBtn"
            id="dailyEnterBtn"
            onClick={[Function]}
          >
            Enter
          </button>
        </div>
        <div>
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="Daily__easterEgg"
  >
    This is an incredibly tiny screen
  </div>
</div>
`;

exports[`Daily rendering renders expected elements 1`] = `
<div>
  <Memo(Connect(ThemeToggle)) />
  <h2
    className="Daily__header"
  >
    Daily Challenge
  </h2>
  <p
    className="Daily__date"
  >
    For the day of 
    
  </p>
  <h3
    className="Daily__words"
  >
    From: 
    
     -&gt; To: 
    
  </h3>
  <div
    className="Daily__timer"
    role="timer"
  />
  <div>
    <div
      aria-labelledby="dailyHistory"
      className="Daily__history"
    >
      <h3
        className="Daily__historyLabel"
        id="dailyHistory"
      >
        History
      </h3>
      <ul
        aria-live="polite"
        className="undefined"
      >
        <li />
      </ul>
    </div>
    <div>
      <HintButton
        ariaLabelledBy="dailyHintHeader"
        btnText="Get a Hint: 3"
        giveSolution={false}
        id="dailyHintBtn"
        isExpanded={false}
        numHints={3}
        onClick={[Function]}
        onExpandChange={[Function]}
      >
        <h3
          id="dailyHintHeader"
        >
          Hint: 
        </h3>
        <p
          className="Daily__hint"
        >
          Recommended next word: 
          
        </p>
        <p
          className="Daily__hint"
        >
          Estimated number of words left: 
          
        </p>
      </HintButton>
      <div
        className="Daily__guessContainer"
      >
        <h3
          className="Daily__guessLabel"
          id="dailyGuessLabel"
        >
          Next Word:
        </h3>
        <div
          className="Daily__guessInputs"
        >
          <div />
          <button
            className="Daily__enterBtn"
            id="dailyEnterBtn"
            onClick={[Function]}
          >
            Enter
          </button>
        </div>
        <div>
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="Daily__easterEgg"
  >
    This is an incredibly tiny screen
  </div>
</div>
`;
